File: repos/remodex-opencode/CodexMobile/CodexMobile/Models/ProviderCapabilities.swift
// FILE: ProviderCapabilities.swift
// Purpose: Capability flags used by the composer to hide/show rows based on
//          the selected model's provider rather than hardcoded identity checks.
// Layer: Model
// Exports: ProviderCapabilities
// Depends on: Foundation

import Foundation

struct ProviderCapabilities: Codable, Hashable, Sendable {
    let supportsAgentSelection: Bool
    let supportsReasoningEffort: Bool
    let supportsFastMode: Bool
    let supportsPlanMode: Bool
    let supportsStreamingTools: Bool
    let supportsApprovals: Bool
    let supportsFork: Bool
    let supportsVoice: Bool
    let supportsDesktopHandoff: Bool
    let supportsSlashCommands: Bool
    let supportsSlashCommandExecute: Bool
    let supportsMCP: Bool
    let supportsWorktree: Bool
    let supportsSkillAutocomplete: Bool
    let supportsStructuredSkillInput: Bool
    let supportsSkillFileInjection: Bool
    let supportsImageAttachments: Bool
    let supportsSteer: Bool
    let supportsQueue: Bool
    let supportsAccessMode: Bool

    init(
        supportsAgentSelection: Bool,
        supportsReasoningEffort: Bool,
        supportsFastMode: Bool,
        supportsPlanMode: Bool,
        supportsStreamingTools: Bool,
        supportsApprovals: Bool,
        supportsFork: Bool,
        supportsVoice: Bool,
        supportsDesktopHandoff: Bool,
        supportsSlashCommands: Bool,
        supportsSlashCommandExecute: Bool = false,
        supportsMCP: Bool,
        supportsWorktree: Bool,
        supportsSkillAutocomplete: Bool,
        supportsStructuredSkillInput: Bool = false,
        supportsSkillFileInjection: Bool = false,
        supportsImageAttachments: Bool = false,
        supportsSteer: Bool,
        supportsQueue: Bool,
        supportsAccessMode: Bool = false
    ) {
        self.supportsAgentSelection = supportsAgentSelection
        self.supportsReasoningEffort = supportsReasoningEffort
        self.supportsFastMode = supportsFastMode
        self.supportsPlanMode = supportsPlanMode
        self.supportsStreamingTools = supportsStreamingTools
        self.supportsApprovals = supportsApprovals
        self.supportsFork = supportsFork
        self.supportsVoice = supportsVoice
        self.supportsDesktopHandoff = supportsDesktopHandoff
        self.supportsSlashCommands = supportsSlashCommands
        self.supportsSlashCommandExecute = supportsSlashCommandExecute
        self.supportsMCP = supportsMCP
        self.supportsWorktree = supportsWorktree
        self.supportsSkillAutocomplete = supportsSkillAutocomplete
        self.supportsStructuredSkillInput = supportsStructuredSkillInput
        self.supportsSkillFileInjection = supportsSkillFileInjection
        self.supportsImageAttachments = supportsImageAttachments
        self.supportsSteer = supportsSteer
        self.supportsQueue = supportsQueue
        self.supportsAccessMode = supportsAccessMode
    }

    enum CodingKeys: String, CodingKey {
        case supportsAgentSelection
        case supportsReasoningEffort
        case supportsFastMode
        case supportsPlanMode
        case supportsStreamingTools
        case supportsApprovals
        case supportsFork
        case supportsVoice
        case supportsDesktopHandoff
        case supportsSlashCommands
        case supportsSlashCommandExecute
        case supportsMCP
        case supportsWorktree
        case supportsSkillAutocomplete
        case supportsStructuredSkillInput
        case supportsSkillFileInjection
        case supportsImageAttachments
        case supportsSteer
        case supportsQueue
        case supportsAccessMode
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        supportsAgentSelection = (try? container.decodeIfPresent(Bool.self, forKey: .supportsAgentSelection)) ?? false
        supportsReasoningEffort = (try? container.decodeIfPresent(Bool.self, forKey: .supportsReasoningEffort)) ?? false
        supportsFastMode = (try? container.decodeIfPresent(Bool.self, forKey: .supportsFastMode)) ?? true
        supportsPlanMode = (try? container.decodeIfPresent(Bool.self, forKey: .supportsPlanMode)) ?? false
        supportsStreamingTools = (try? container.decodeIfPresent(Bool.self, forKey: .supportsStreamingTools)) ?? true
        supportsApprovals = (try? container.decodeIfPresent(Bool.self, forKey: .supportsApprovals)) ?? true
        supportsFork = (try? container.decodeIfPresent(Bool.self, forKey: .supportsFork)) ?? true
        supportsVoice = (try? container.decodeIfPresent(Bool.self, forKey: .supportsVoice)) ?? false
        supportsDesktopHandoff = (try? container.decodeIfPresent(Bool.self, forKey: .supportsDesktopHandoff)) ?? false
        supportsSlashCommands = (try? container.decodeIfPresent(Bool.self, forKey: .supportsSlashCommands)) ?? true
        supportsSlashCommandExecute =
            (try? container.decodeIfPresent(Bool.self, forKey: .supportsSlashCommandExecute)) ?? false
        supportsMCP = (try? container.decodeIfPresent(Bool.self, forKey: .supportsMCP)) ?? false
        supportsWorktree = (try? container.decodeIfPresent(Bool.self, forKey: .supportsWorktree)) ?? false
        supportsSkillAutocomplete = (try? container.decodeIfPresent(Bool.self, forKey: .supportsSkillAutocomplete)) ?? false
        supportsStructuredSkillInput =
            (try? container.decodeIfPresent(Bool.self, forKey: .supportsStructuredSkillInput)) ?? false
        supportsSkillFileInjection =
            (try? container.decodeIfPresent(Bool.self, forKey: .supportsSkillFileInjection)) ?? false
        supportsImageAttachments =
            (try? container.decodeIfPresent(Bool.self, forKey: .supportsImageAttachments)) ?? false
        supportsSteer = (try? container.decodeIfPresent(Bool.self, forKey: .supportsSteer)) ?? false
        supportsQueue = (try? container.decodeIfPresent(Bool.self, forKey: .supportsQueue)) ?? true
        supportsAccessMode = (try? container.decodeIfPresent(Bool.self, forKey: .supportsAccessMode)) ?? true
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(supportsAgentSelection, forKey: .supportsAgentSelection)
        try container.encode(supportsReasoningEffort, forKey: .supportsReasoningEffort)
        try container.encode(supportsFastMode, forKey: .supportsFastMode)
        try container.encode(supportsPlanMode, forKey: .supportsPlanMode)
        try container.encode(supportsStreamingTools, forKey: .supportsStreamingTools)
        try container.encode(supportsApprovals, forKey: .supportsApprovals)
        try container.encode(supportsFork, forKey: .supportsFork)
        try container.encode(supportsVoice, forKey: .supportsVoice)
        try container.encode(supportsDesktopHandoff, forKey: .supportsDesktopHandoff)
        try container.encode(supportsSlashCommands, forKey: .supportsSlashCommands)
        try container.encode(supportsSlashCommandExecute, forKey: .supportsSlashCommandExecute)
        try container.encode(supportsMCP, forKey: .supportsMCP)
        try container.encode(supportsWorktree, forKey: .supportsWorktree)
        try container.encode(supportsSkillAutocomplete, forKey: .supportsSkillAutocomplete)
        try container.encode(supportsStructuredSkillInput, forKey: .supportsStructuredSkillInput)
        try container.encode(supportsSkillFileInjection, forKey: .supportsSkillFileInjection)
        try container.encode(supportsImageAttachments, forKey: .supportsImageAttachments)
        try container.encode(supportsSteer, forKey: .supportsSteer)
        try container.encode(supportsQueue, forKey: .supportsQueue)
        try container.encode(supportsAccessMode, forKey: .supportsAccessMode)
    }

    static let defaultCodex = ProviderCapabilities(
        supportsAgentSelection: false,
        supportsReasoningEffort: true,
        supportsFastMode: true,
        supportsPlanMode: true,
        supportsStreamingTools: true,
        supportsApprovals: true,
        supportsFork: true,
        supportsVoice: true,
        supportsDesktopHandoff: true,
        supportsSlashCommands: true,
        supportsSlashCommandExecute: false,
        supportsMCP: true,
        supportsWorktree: true,
        supportsSkillAutocomplete: true,
        supportsStructuredSkillInput: true,
        supportsSkillFileInjection: true,
        supportsImageAttachments: true,
        supportsSteer: true,
        supportsQueue: true,
        supportsAccessMode: true
    )

    static let defaultOpenCode = ProviderCapabilities(
        supportsAgentSelection: true,
        supportsReasoningEffort: false,
        supportsFastMode: false,
        supportsPlanMode: false,
        supportsStreamingTools: true,
        supportsApprovals: true,
        supportsFork: true,
        supportsVoice: false,
        supportsDesktopHandoff: false,
        supportsSlashCommands: true,
        supportsSlashCommandExecute: true,
        supportsMCP: false,
        supportsWorktree: false,
        supportsSkillAutocomplete: true,
        supportsStructuredSkillInput: false,
        supportsSkillFileInjection: true,
        supportsImageAttachments: true,
        supportsSteer: false,
        supportsQueue: true,
        supportsAccessMode: false
    )
}

File: repos/remodex-opencode/phodex-bridge/src/provider-capabilities.js
// FILE: provider-capabilities.js
// Purpose: Defines capability flags for runtimes. Used by runtime/catalog and model/list
//          to drive iOS composer visibility/grey-out without hardcoded provider checks.
// Layer: Bridge runtime utility
// Exports: CAPABILITIES, resolveModelCapabilities, resolveOpenCodeCatalogCapabilities, hasReasoningEffort
// Depends on: ./opencode-models

const { CODEX_PROVIDER_ID, OPENCODE_PROVIDER_ID } = require("./opencode-models");


const CAPABILITIES = [
  "supportsAgentSelection",
  "supportsReasoningEffort",
  "supportsFastMode",
  "supportsPlanMode",
  "supportsVoice",
  "supportsDesktopHandoff",
  "supportsWorktree",
  "supportsFork",
  "supportsApprovals",
  "supportsStreamingTools",
  "supportsSlashCommands",
  "supportsSlashCommandExecute",
  "supportsMCP",
  "supportsSkillAutocomplete",
  "supportsStructuredSkillInput",
  "supportsSkillFileInjection",
  "supportsImageAttachments",
  "supportsSteer",
  "supportsQueue",
];

const CODEX_CAPABILITIES = {
  supportsAgentSelection: false,
  supportsReasoningEffort: true,
  supportsFastMode: true,
  supportsPlanMode: true,
  supportsVoice: true,
  supportsDesktopHandoff: true,
  supportsWorktree: true,
  supportsFork: true,
  supportsApprovals: true,
  supportsStreamingTools: true,
  supportsSlashCommands: true,
  supportsSlashCommandExecute: false,
  supportsMCP: true,
  supportsSkillAutocomplete: true,
  supportsStructuredSkillInput: true,
  supportsSkillFileInjection: true,
  supportsImageAttachments: true,
  supportsSteer: true,
  supportsQueue: true,
};

const OPENCODE_CAPABILITIES = {
  supportsAgentSelection: true,
  supportsReasoningEffort: false,
  supportsFastMode: false,
  supportsPlanMode: false,
  supportsVoice: false,
  supportsDesktopHandoff: true,
  supportsWorktree: false,
  supportsFork: true,
  supportsApprovals: true,
  supportsStreamingTools: true,
  supportsSlashCommands: true,
  supportsSlashCommandExecute: true,
  supportsMCP: false,
  supportsSkillAutocomplete: true,
  // RP-SKILL-3: kept false pending upstream SDK support for skills:[] (or dedicated skill parts) in
  // session.prompt / V2 prompt input. See opencode-sdk.md for shapes inspected (PromptInput uses only
  // parts: Text|File|Agent|Subtask; V2 Prompt uses text+files/agents/references; no skills array).
  // Bridge now wires conditional skills[] (gated); iOS sends structured items only when flag true.
  // Do not force-enable; catalog test + DISABLE regression cover. Flip only after SDK + device E2E.
  supportsStructuredSkillInput: false,
  supportsSkillFileInjection: true,
  supportsImageAttachments: true,
  supportsSteer: false,
  supportsQueue: true,
};

function resolveOpenCodeCatalogCapabilities(_env = process.env) {
  return { ...OPENCODE_CAPABILITIES };
}

function resolveModelCapabilities(providerId, modelData = {}, env = process.env) {
  modelData = modelData || {};
  const base =
    providerId === CODEX_PROVIDER_ID
      ? { ...CODEX_CAPABILITIES }
      : providerId === OPENCODE_PROVIDER_ID
        ? resolveOpenCodeCatalogCapabilities(env)
        : { ...CODEX_CAPABILITIES };

  if (Array.isArray(modelData.supportedReasoningEfforts)) {
    base.supportsReasoningEffort = modelData.supportedReasoningEfforts.length > 0;
  }

  const reasoningEffortsAlt = Array.isArray(modelData.reasoningEfforts)
    ? modelData.reasoningEfforts
    : [];
  if (reasoningEffortsAlt.length > 0) {
    base.supportsReasoningEffort = true;
  }

  if (modelData.supportsFastMode === true) {
    base.supportsFastMode = true;
  }

  if (modelData.supportsReasoning !== undefined) {
    base.supportsReasoningEffort = Boolean(modelData.supportsReasoning);
  }

  if (modelData.hasReasoning !== undefined) {
    base.supportsReasoningEffort = Boolean(modelData.hasReasoning);
  }

  return base;
}

function hasReasoningEffort(capabilities) {
  return capabilities?.supportsReasoningEffort === true;
}

module.exports = {
  CAPABILITIES,
  CODEX_CAPABILITIES,
  OPENCODE_CAPABILITIES,
  resolveOpenCodeCatalogCapabilities,
  resolveModelCapabilities,
  hasReasoningEffort,
};
